query-languages/m/m-spec-types.md

```diff
@@ -353,7 +353,7 @@ Value.Type( Value.ReplaceType( {1}, type {number} )
 
 ## Type equivalence and compatibility
 
-Type equivalence is not defined in M. Any two type values that are compared for equality may or may not return `true`. However, the relation between those two types (whether `true` or `false`) will always be the same.
+Type equivalence is not fully defined in M. It is acceptable for an M implementation to return `false` when type values are compared for equality. If `true` is returned, at minimum, the following criteria must be met: the two values must be based on the same nullable primitive type and the response returned must be consistent if the same values are repeatedly compared (that is, multiple recurrences of the same comparison must all either evaluate to `true` or to `false`). 
```

Comment:
How about simplifying to the following (which I think was the original intent)?

```suggestion
Type equivalence is not defined in M. Two identical type values that are compared for equality may or may not return `true`. However, the equality of those two types (whether `true` or `false`) will always be the same.
```

Reply:
The problem with explaining the details of how the implementation works is summed up by the first sentence:
_Type equivalence is not defined in M._

Reply:
To me, the statement that "two identical values...may or may not return true" is a bit confusing to make sense out of. Having worked with (and researched) M for quite a while, I believe I understand the meaning of it, but when I first read similar phraseology in the spec years ago, exactly what was going on was unclear.

What do you think of the revision I just committed in an attempt to state this more clearly using different words?

Reply:
I think it's clearer than what you wrote, even though the behavior it's describing can be confusing.

To give an example:

(type text) = (type text)
returns true

(type [a = text]) = (type [a = text])
returns false

Two types can be identical, but they still may not compare equal in M.

The new wording in your latest iteration makes it sound like two completely dissimilar types can compare equal. I don't think that's the intent here.

Reply:
> The new wording in your latest iteration makes it sound like two completely dissimilar types can compare equal. I don't think that's the intent here.

Agreed that this is not the intent. However, isn't this also what the current wording implies by its "any...may or may not return true"?

Is the below a fair summary of the spec's perspective on type equality comparisons (not saying this should necessarily go verbatim into the spec, but to make sure I'm following on the same page with you on how things work):

* The M language does not define type equivalence rules.
* An M implementation may **optionally** choose to implement support for type equivalence comparisons (including for only a **subset** of type values). When a type equality comparison is attempted that is not supported by the implementation, `false` should be returned.
* When an M implementation supports comparing two particular type values for equality:
  * If the two values could reasonably be considered equivalent, `true` **may** be returned (but the exact specifics of how this equivalency is determined is left to the discretion of the implementation). Otherwise, `false` must be returned.
  * Multiple recurrences of comparing the same two type value instances must consistently either evaluate to `true` or to `false`.

Reply:
Yes, your summary seems accurate.